src/ripple/app/consensus/RCLCensorshipDetector.h

```diff
@@ -0,0 +1,127 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_CONSENSUS_RCLCENSORSHIPDETECTOR_H_INCLUDED
+#define RIPPLE_APP_CONSENSUS_RCLCENSORSHIPDETECTOR_H_INCLUDED
+
+#include <ripple/shamap/SHAMap.h>
+#include <algorithm>
+#include <map>
+
+namespace ripple {
+
+template <class TxID, class Sequence>
+class RCLCensorshipDetector
+{
+private:
+    std::map<TxID, Sequence> tracker;
+
+    /** Removes all elements satisfying specific criteria from the tracker
+
+        @param pred A predicate which returns true for tracking entries that
+                    should be removed. The predicate must be callable as:
+                        bool pred(TxID const&, Sequence)
+                    It must return true for entries that should be removed.
+    */
+    template <class Predicate>
+    void prune(Predicate pred)
+    {
+        auto t = tracker.begin();
+
+        while (t != tracker.end())
+        {
+            if (pred(t->first, t->second))
+                t = tracker.erase(t);
+            else
+                t = std::next(t);
+        }
+    }
+
+public:
+    RCLCensorshipDetector() = default;
+
+    /** Add transactions being proposed for the current consensus round.
+
+        @param seq      The sequence number of the ledger being built.
+        @param proposed The set of transactions that we are initially proposing
+                        for this round.
+    */
+    void propose(
+        Sequence seq,
+        std::vector<TxID> proposed)
```

Comment:
The reason `proposed` is passed by value is because of the `sort`. This complicated the interface because callers need to `std::move` the parameter to avoid the copy (if they don't use the collection again) or need to copy if they do - even if the collection was already sorted.

I honestly don't have a good solution. `boost::flat_set` is the best I can come up with. However, that has quadradic construction, and it's much more likely callers will already have a `std::vector`. Since N is small the extra construction doesn't bug me too much. If you don't like that (and I'm on the fence), then we at least need to change the callers to move this parameters. I will leave comments at the call sites.